aspnet/fundamentals/localization.rst

```diff
@@ -1,8 +1,285 @@
+
+Use ``RequestLocalizationOptions`` to add or remove localization providers. 
+
+Suppose you want to let your customers store their language and culture in your databases. You could write a provider to look up these values for the user. The following code shows how to add a custom provider:
+
+.. code-block:: c#
+
+    services.Configure<RequestLocalizationOptions>(options =>
+    {
+        var supportedCultures = new[]
+        {
+            new CultureInfo("en-US"),
+            new CultureInfo("fr")
+        };
+
+        options.DefaultRequestCulture = new RequestCulture(culture: "en-US", uiCulture: "en-US");
+        options.SupportedCultures = supportedCultures;
+        options.SupportedUICultures = supportedCultures;
+
+        options.RequestCultureProviders.Insert(0, new CustomRequestCultureProvider(async context =>
+        {
+           // My custom request culture logic
+          return new ProviderCultureResult("en");
+        }));
+    });
+
+Configuring localization
+----------------------------------
+
+Localization is configured in the ``ConfigureServices`` method:
+
+.. literalinclude:: localization/sample/StartupConfig.cs
+  :language: c#
+  :linenos:
+  :lines: 44-45,55-61
+  :dedent: 6
+
+- ``AddLocalization`` Adds the localization services to the services container. The code above also sets the resources path to "Resources".
+- ``AddViewLocalization`` Adds support for localized view files. In this sample view localization is based on the view file suffix. For example "fr" in the *Index.fr.cshtml* file.
+- ``AddDataAnnotationsLocalization`` Adds support for localized ``DataAnnotations`` validation messages through ``IStringLocalizer`` abstractions.
+
+
+Resource file naming
+---------------------
+
+When you create a new project with Visual Studio, the project name is used for the assembly name, project name and default namespace. For example, if you created a project named **X.Y.Z**, the assembly name, project name and default namespace would all be "X.Y.Z".
+
+.. image:: localization/_static/newproj.png
+
+When you create a type in a subfolder, Visual Studio appends the folder name to the default namespace (it nests the namespace). C# uses the "." between nested namespaces and types. Using the **X.Y.Z** project, a class named ``Cat`` in the *Models* folder would have the namespace ``X.Y.Z.Models``:
+
+.. code-block:: c# 
+
+    namespace X.Y.Z.Models
+    {
+        class Cat
+        {
+            public string name { get; set; }
+        }
+    }
+
+If you reference a type with the fully qualified namespace (longer than the default namespace), Visual Studio will warn you to use the nested or *simplified namespace*:
+
+.. image:: localization/_static/simp2.png
+
+In the image above, Visual Studio is warning you that ``Models.Cat`` can be used for the type, because ``Models.Cat`` is in scope. The code below shows the ``Cat`` class with both the full namespace and the nested namespace.
```

Comment:
@danroth27 I can remove it if you like, but defining the exact naming convention is trickery than you might think. This is also the biggest problem customers have been facing. Because the VS IDE does all the namespace work for you, many customers don't realize what the nested namespace is - they just know what the IDE generates and accepts. If you get the NS wrong, the IDE provides instant feedback. With resource files, it's not until run-time and then you're not sure why the resource file is not read.